app/gui/src/project-view/providers/interactionHandler.ts

```diff
@@ -44,37 +70,70 @@ export class InteractionHandler {
     if (this.isActive(interaction)) this.currentInteraction.value = undefined
   }
 
-  /** End the current interaction, if it is the specified instance. */
+  /**
+   * End the interaction, if it is currently active.
+   * Any children interactions of the given interaction will be ended as well.
+   */
   end(interaction: Interaction) {
-    if (this.isActive(interaction)) {
-      this.currentInteraction.value = undefined
-      interaction.end?.()
+    search(this.currentInteraction.value, interaction, (found, children) => {
+      this.currentInteraction.value = found.parentInteraction
+      found.end?.()
+      for (const interaction of children) {
+        interaction.end?.()
+      }
+    })
+  }
+
+  /** End all interactions, in the order from child to parent. */
+  endAll() {
+    let current = this.currentInteraction.value
+    while (current != null) {
+      current.end?.()
+      current = current.parentInteraction
     }
+    this.currentInteraction.value = undefined
   }
 
-  /** Cancel the current interaction, if it is the specified instance. */
+  /**
+   * Cancel the interaction, if it is currently active.
+   * Any children interactions of the given interaction will be cancelled as well.
+   */
   cancel(interaction: Interaction) {
-    if (this.isActive(interaction)) {
-      this.currentInteraction.value = undefined
-      interaction.cancel?.()
-    }
+    search(this.currentInteraction.value, interaction, (found, children) => {
+      this.currentInteraction.value = found.parentInteraction
+      found.cancel?.()
+      for (const interaction of children) {
+        interaction.cancel?.()
+      }
+    })
   }
 
-  /** TODO: Add docs */
-  handleCancel(): boolean {
+  /**
+   * Cancel all interactions, in the order from child to parent.
+   * @returns `true` if the current interaction was cancelled.
+   */
+  cancelAll(): boolean {
     const hasCurrent = this.currentInteraction.value != null
-    this.currentInteraction.value?.cancel?.()
+    let interaction = this.currentInteraction.value
+    while (interaction != null) {
+      interaction.cancel?.()
+      interaction = interaction.parentInteraction
+    }
     this.currentInteraction.value = undefined
     return hasCurrent
   }
 
   /**
-   * Handle pointer event in capture. Calls `pointerdown` handler of currently active handler.
+   * Handle pointer event in capture. Calls the corresponding handler of the currently active interaction.
    *
    * Because usually the handlers check for clicks outside the active panel, even if event is handled,
    * it is NOT stopped, and its default action is NOT prevented.
+   *
+   * Only the current interaction is considered, its ancestors are not notified.
    */
-  handlePointerEvent<HandlerName extends keyof Interaction>(
+  handlePointerEvent<
+    HandlerName extends keyof Omit<Interaction, 'cancel' | 'end' | 'parentInteraction'>,
+  >(
```

Comment:
Do we need to omit? I see the non-event-handler interaction fields are checked in line 138.

Reply:
Yeah, it is needed, because the check on line 138 was not working correctly (you could set up a handler for `cancel`, for example). Turning `handler` type to `never` does not prevent you from calling the function, unlike the restriction on the generic. Thinking about it now, I will remove all this typescript magic and leave only `'pointerup' | 'pointerdown'` type constraint.